File: solutions/python/resistor-color-trio/resistor_color_trio.py
"""Parse resistor colors."""

import enum


class Colors(enum.IntEnum):
    """Color mapping."""
    black = 0
    brown = 1
    red = 2
    orange = 3
    yellow = 4
    green = 5
    blue = 6
    violet = 7
    grey = 8
    white = 9


def label(colors: list[str]) -> str:
    """Return the value of a resistor color."""
    val = 0
    for color in colors[:2]:
        val = val * 10 + Colors[color].value
    val *= 10 ** Colors[colors[2]].value
    if val % 1000 == 0:
        return f"{val // 1000} kiloohms"
    return f"{val} ohms"

File: solutions/ruby/resistor-color-trio/resistor_color_trio.rb
class ResistorColorTrio
  private

  COLORS = %w[black brown red orange yellow green blue violet grey white].freeze
  UNITS = %w[ohms kiloohms megaohms gigaohms].freeze

  attr_writer :label

  def initialize(colors)
    tens, ones, power = colors
    base = COLORS.index(tens) * 10 + COLORS.index(ones)
    resistance = base * 10**COLORS.index(power)

    shifts = 0
    while !resistance.zero? && (resistance % 1000).zero?
      shifts += 1
      resistance /= 1000
    end
    self.label = "Resistor value: #{resistance} #{UNITS[shifts]}"
  end

  public

  attr_reader :label
end
